File: Trees/BalancedBinaryTree.cpp
// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    bool isBalanced(TreeNode *root)
    {
        bool cond = true;
        dfs(root, cond);
        return cond;
    }

    int dfs(TreeNode *root, bool &cond)
    {
        if (root == nullptr)
        {
            return 0;
        }

        int l = dfs(root->left, cond);
        int r = dfs(root->right, cond);

        if (abs(l - r) > 1)
        {
            cond = false;
        }
        return 1 + max(l, r);
    }
};

File: rust/src/trees/balanced_binary_tree.rs
use std::rc::Rc;
use std::cell::RefCell;

use crate::trees::{Solution, TreeNode};

impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let mut cond = true;
        Self::height(root, &mut cond);
        cond
    }

    pub fn height(curr: Option<Rc<RefCell<TreeNode>>>, cond: &mut bool) -> i32 {
        if let Some(value) = curr {
            let left_height = Self::height(value.borrow().left.clone(), cond);
            let right_height = Self::height(value.borrow().right.clone(), cond);

            if (left_height - right_height).abs() > 1 {
                *cond = false;
            }

            return 1 + std::cmp::max(left_height, right_height);
        }
        0
    }
}